plugins/ansilexer.py

```diff
@@ -28,83 +29,288 @@
-            ('\x1b\\[([01])m\x1b\\[(\\d+)m([^\x1b]*)', callback),
-            ('\x1b\\[(\\d+)(?:;(\\d+))?m([^\x1b]*)', callback)
+            (_sgrs_and_text, _callback),
         ]
     }
 
-_ansi_fg_color_re = re.compile('class="g g-AnsiForegroundColor([0-9a-f]{6})">')
-_ansi_fg_bg_color_re = re.compile('class="g g-AnsiForegroundBackgroundColor([0-9a-f]{6})">')
-
 class HtmlAnsiFormatter(HtmlFormatter):
+    _ansi_color_re = re.compile(
+        r'(?P<Prefix>class="g )(?P<AnsiClass>g-Ansi\S+)(?P<Suffix>">)'
+    )
+    _rgb_code_re = re.compile(r'[0-9a-f]{6}$')
+
     def wrap(self, source, outfile):
         return self._wrap_code(source)
 
     def _wrap_code(self, source):
         for i, t in source:
             if i == 1: # it's a line of formatted code
-                # Add ZWNJ for before each character because otherwise it's
-                # somehow impossible to wrap even with word-break: break-all.
-                # Not sure why (and not sure if this is the best solution), but
-                # had to ship a thing so there it is. Adding <wbr/> did not
-                # help.
-                t = _ansi_fg_bg_color_re.sub('style="color: #\\1; background-color: #\\1">&zwnj;', t)
-                t = _ansi_fg_color_re.sub('style="color: #\\1">&zwnj;', t)
-                #t += 'H'
+                t = self._ansi_color_re.sub(self._replace_ansi_class, t)
             yield i, t
+
+    def _replace_ansi_class(self, match):
+        html_classes = ['g']
+        html_styles = []
+        foreground_color = None
+        background_color = None
+        ansi_class = match.group('AnsiClass')
+
+        i = ansi_class.find('BackgroundColor')
+        if i >= 0:
+            background_color = ansi_class[i + 15:]
+            ansi_class = ansi_class[:i]
+
+            if self._rgb_code_re.match(background_color):
+                html_styles.append('background-color: #' + background_color)
+            else:
+                html_classes.append('g-AnsiBackground' + background_color)
+
+        i = ansi_class.find('ForegroundColor')
+        if i >= 0:
+            foreground_color = ansi_class[i + 15:]
+            ansi_class = ansi_class[:i]
+            html_styles.insert(0, 'color: #' + foreground_color)
+
+        if ansi_class != 'g-Ansi':
+            html_classes.insert(1, ansi_class)
+
+        result = ''
+        if len(html_classes) > 1: # we don't want to emit just g
+            result = 'class="' + ' '.join(html_classes) + '"'
+
+        if html_styles:
+            if result:
+                result += ' '
+            result += 'style="' + '; '.join(html_styles) + '"'
+
+        # Add ZWNJ at the start of each span because otherwise it's
+        # somehow impossible to wrap even with word-break: break-all.
+        # Not sure why (and not sure if this is the best solution), but
+        # had to ship a thing so there it is. Adding <wbr/> did not
+        # help.
+        result += '>&zwnj;'
```

Comment:
I'll admit I didn't understand this stuff at all. You'd been inserting it only when emitting RGB-defined colors, and I... didn't really know what to do, so I guessed, and am emitting everywhere. Happy to change.

Reply:
Yeah, sorry. This was to make the wrapped terminal output work on this post: https://blog.magnum.graphics/backstage/multidimensional-strided-array-views/ -- without the `&zwnj;`  the browser didn't really want to wrap the output for some reason, putting scrollbars on the thing, and ugh.

I'm adding "hooks" for code block postprocessing so I guess the ZWNJ isn't really needed anymore and I'll do it on my side instead. Can you drop it, then? Thanks.